src/libs/Notification/LocalNotification/BrowserNotifications.ts

```diff
@@ -44,19 +44,24 @@ function canUseBrowserNotifications(): Promise<boolean> {
  * @param icon Path to icon
  * @param data extra data to attach to the notification
  */
-function push(title: string, body = '', icon: string | ImageSourcePropType = '', data: LocalNotificationData = {}, onClick: LocalNotificationClickHandler = () => {}, silent = false) {
+function push(title: string, body = '', icon: string | ImageSourcePropType = '', data: LocalNotificationData = {}, onClick: LocalNotificationClickHandler = () => {}, silent = false, tag = '') {
     canUseBrowserNotifications().then((canUseNotifications) => {
         if (!canUseNotifications) {
             return;
         }
 
+        // We want to show only one instance of notifications having the same tag
```

Comment:
```suggestion
        // We want to show only one instance of notifications having the same tag
        if (tag && Object.values(notificationCache).some((notification) => notification.tag === tag)) {
            return
        }
```

Reply:
I had tried that previously. `tag` only groups the notifications together but it still shows the duplicate notifications unless we use a custom logic. Even the document only mentioned about grouping. What do you think?

Reply:
Just tried again on Mac desktop, only passing tag still leads to multiple notifications.

Reply:
@ShridharGoel Agree with you. It will be more convincing to link a reference about this issue on the Desktop. But I can't find any reference. 😄

Reply:
@DylanDylann I couldn't find any reference as well.

Reply:
Ah okay so it doesn't work correctly on desktop? If so, let's keep this logic but change it to replace the existing notification like browsers should. Also add a comment to explain why it's needed for desktop

Reply:
@arosiclair `close()` method is not working for desktop. Any suggestion on how to dismiss the previous notification before adding the new one?

Reply:
```
if (tag) {
            const notificationId = Object.keys(notificationCache).find(notificationID => {
                const notification = notificationCache[notificationID]
                return notification.tag === tag
            })
            if (notificationId) {
                notificationCache[notificationId].close()
                delete notificationCache[notificationId]
            }
        }
```

Reply:
@arosiclair 

>  If so, let's keep this logic but change it to replace the existing notification like browsers should

Because we have multiple updates and each update triggers a notification.
We have 2 notification badges that are the same, Why do we need to replace the first one with a new one? It may become flickering

Reply:
> @arosiclair close() method is not working for desktop. Any suggestion on how to dismiss the previous notification before adding the new one?

`close()` should work on desktop. That is what we're using [here](https://github.com/Expensify/App/blob/401bb219c74fb615704895554b6876eb25601a11/src/libs/Notification/LocalNotification/BrowserNotifications.ts#L140). Maybe there is some other bug? Also you do not need to delete the notification from the cache here. That is already handled in the `onclose` handler.

> Why do we need to replace the first one with a new one? It may become flickering

That is the expected behavior when pushing a notification with a `tag`. Our update notifications are identical, but others are not. Flickering shouldn't be much of an issue.

Reply:
> Our update notifications are identical, but others are not

A bit of wondering, currently we will ignore the new notification if its tag matches the tag of another notification. IMO, in every case. If two tags are the same, we will consider them as the same notifications. Do you mean that there will be cases where two notifications have the same tag but the content of the notification is different?

Reply:
On desktop:
Both, simply adding the tag as well as [this](https://github.com/Expensify/App/pull/37595#discussion_r1517832325) code works when a single useEffect is used. But when we are using two of them and creating notifications at the same time, then it starts showing multiple of them. Only the previous code (early return if notification with tag already exists) works there.

Reply:
Basically we can just pass tag if we only want to support such kind of scenario:
```
useEffect(() => {
        const interval = setInterval(() => {
            console.log("Update available 1")
            LocalNotification.showUpdateAvailableNotification();
        }, 5000);
        return () => clearInterval(interval);
    }, [])
```

But we need to prevent creating new notifications by returning if tag matches, if we want to handle this:
```
useEffect(() => {
        const interval = setInterval(() => {
            console.log("Update available 1")
            LocalNotification.showUpdateAvailableNotification();
        }, 5000);
        return () => clearInterval(interval);
    }, [])

    useEffect(() => {
        const interval = setInterval(() => {
            console.log("Update available 2")
            LocalNotification.showUpdateAvailableNotification();
        }, 5000);
        return () => clearInterval(interval);
    }, [])
```

Reply:
Thanks, let's wait @arosiclair confirm [this comment](https://github.com/Expensify/App/pull/37595#discussion_r1517910862) then I will review and test your new change

Reply:
> Do you mean that there will be cases where two notifications have the same tag but the content of the notification is different?

Yes, tags are not a unique identifier. They are used to replace old notifications with the same tag. Please see [the docs](https://developer.mozilla.org/en-US/docs/Web/API/Notification/tag).

> But we need to prevent creating new notifications by returning if tag matches, if we want to handle this:

It seems like you are creating a race condition with this? I don't think we need to worry about this at this point. If it does become an issue in real world use, we can probably add some kind of lock to fix it.

Reply:
@arosiclair @DylanDylann Updated, only passing the tag now.

Reply:
If there are many update notifications, we expect to display one of them instead of displaying multiple same notifications
<img width="615" alt="Screenshot 2024-03-09 at 01 37 47" src="https://github.com/Expensify/App/assets/141406735/a840a7ee-3029-46fd-8f71-aab9c7050962">

Please correct me if I am wrong

Reply:
@DylanDylann Yeah, I mentioned this [here](https://github.com/Expensify/App/pull/37595#discussion_r1517964126) but @arosiclair mentioned that we can ignore it (also I think this scenario would be unlikely for update notifications).

Reply:
> also I think this scenario would be unlikely for update notifications

It is maybe an amazing point. But we need to deep dive into it and find out a way to reproduce the original bug on dev. So we can prove that our solution works well

Reply:
The original bug can be seen by just putting this code in Expensify.js

```
useEffect(() => {
        const interval = setInterval(() => {
            console.log("Update available")
            LocalNotification.showUpdateAvailableNotification();
        }, 5000);
        return () => clearInterval(interval);
    }, [])
```

Reply:
Only 1 useEffect instead of 2.

Reply:
Yes let's not make this overly complex. In the real world, updates are only deployed once a day and internally we may do a few staging deploys to fix blockers. There's no need to simulate or fix race conditions for that. Testing once every 5 seconds (one effect like above) is more than enough.

Reply:
@arosiclair Tested on both web and desktop.